docs/proposals/message-proto-design.md

```diff
@@ -0,0 +1,346 @@
+
+# change model message to  protobuf message 
+
+## Abstract 
+
+currently, In beehive framework, model message is defined and serialized using json format. the performance is slower than protobuf.
+
+and for parsing the message resource, each field index is hardcoded. such as (https://github.com/kubeedge/beehive/blob/master/pkg/common/util/parse_resource.go)
+
+in kubeedge project , we can see mush `Split(string, "/")` function which to parse the resource path string.  these are hard code.   when we deal will the message resource after we receive message  . we need to know the format of the assembly In advance。 then use `Split()` function to get the field we want.    it`s not easy to expand. It's not so extensible. so we need a more extended form.
+
+please see the #399 issue of kubeeedge 
+
+in addition, currently  when we use quic or websocket protocol to paas message, we need convert model message to probuf message (which defined in `github.com/kubeedge/viaduct/pkg/protos/message/message.pb.go`). and then send .  This also affects the efficiency of the transmission, rather than simply turning the model mesage into proto message。
+
+## Need to do
+
+*  Change the model message to protobuf message  directly
+*  Solve the filed index problem of `Resource`
+*  Consider compatibility  
+
+## How to do 
+
+##  Change the model message to protobuf message  directly
+
+create proto file in `model` dir to define message struct. and use `go generate`(will move to makefile) to create message.pb.go file 
+
+```
+syntax = "proto3";
+package model;
+import "google/protobuf/any.proto";
+
+message MessageHeader {
+    string ID = 1;
+    string ParentID = 2;
+    int64 Timestamp = 3;
+    bool Sync = 4;
+}
+message MessageRoute {
+    string Source = 1;
+    string Group = 2;
+    string Operation = 3;
+    google.protobuf.Any Resource = 4;
```

Comment:
can't this be of type string ?

Reply:
can't this be of type bytes?
```suggestion
    bytes Resource = 4;
```

Reply:
@rohitsardesai83  if we use string type, we need use `Split(string, "/")` function to parse the resource path string.  these are so many hard code.

Reply:
@lzx258  bytes types is ok, but if we use bytes, We don't know what type to convert bytes to。  `google.protobuf.Any` contains 
```
 string type_url = 1;

  // Must be a valid serialized protocol buffer of the above specified type.
  bytes value = 2;
}
```
it also contains an arbitrary serialized message as bytes. `Any` message type lets you use messages as embedded types without having their .proto definition.